Add `GeneratedAsIdentityType` enum attribute to column definition

ZhouXing19 opened the pull request:
This commit add a enum attribute to note how the column creation is 
associated with the `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` token.
If a column is created with `GENERATED ALWAYS AS IDENTITY`, 
it is not allowed to be overridden without  `OVERRIDING SYSTEM VALUE` token 
in the `INSERT | UPSERT | UPDATE` statement.
See also #68201

We implement `GeneratedAsIdentityType` enum following this logic:

1. Persistent column definition: Add `enum GeneratedAsIdentityType`, 
and add `GeneratedAsIdentityType generated_as_id_type` 
under `ColumnDescriptor`
at `pkg/sql/catalog/descpb/structured.proto`

2. Write access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
set the variable when `GENERATED ALWAYS AS IDENTITY` is in the
`CREATE TABLE` statement. 
Check `pkg/sql/catalog/tabledesc/table.go`

3. Read access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
refactored `cat.Column.Init()` at `pkg/sql/opt/cat/column.go`

4. Check if a column is created with 
`GENERATED ALWAYS AS IDENTITY` when
insert: add `func checkColumnIsGeneratedAlways`
at `pkg/sql/opt/optbuilder/mutation_builder.go`
and implement it in `pkg/sql/opt/optbuilder/insert.go`,
`pkg/sql/opt/optbuilder/insert.go`
, and `pkg/sql/opt/optbuilder/update.go`

We also added logic tests at
`pkg/sql/logictest/testdata/logic_test/alter_table`
and 
`pkg/sql/logictest/testdata/logic_test/create_table`

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68324)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
this is looking good!

i had a few naming and error handling comments

also, for the commit message, we can use `Release note: none` for this, since this change is not visible to any users

i'm less familiar with the optimizer code. would you be able to find someone from the SQL Queries team to review this? i'll suggest @mgartner since he recently added "inaccessible" columns which seem fairly similar in where they need to be checked.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-Mg78r9b9AhxznBRMRJj:b-hegltb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75k-349X25q6NjA5r:-Mg75k-349X25q6NjA5s:byk39bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L151)):*
> ```protobuf
>   optional bool inaccessible = 17 [(gogoproto.nullable) = false];
> 
>   // A disallowexplicitwrit column does not allow explicit insert.
> ```

nit: fix spelling in the comment
___
*[pkg/sql/catalog/descpb/structured.proto, line 153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg76Ada5WgCmEE5tiTw:-Mg76Ada5WgCmEE5tiTx:b-ex0ak0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L153)):*
> ```protobuf
>   // A disallowexplicitwrit column does not allow explicit insert.
>   // To insert a disallowexplicitwrit column, please specify the OVERRIDING SYSTEM VALUE
>   // token in the INSERT statement.
> ```

nit: instead of "please specify the OVERRIDING SYSTEM VALUE token in the INSERT statement"

can it say "the INSERT statement must include the OVERRIDING SYSTEM VALUE clause"
___
*[pkg/sql/catalog/descpb/structured.proto, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Z3q9LVcgDfht2S6:-Mg75Z3q9LVcgDfht2S7:b8ywuoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L154)):*
> ```protobuf
>   // To insert a disallowexplicitwrit column, please specify the OVERRIDING SYSTEM VALUE
>   // token in the INSERT statement.
>   optional bool disallowexplicitwrite = 18 [(gogoproto.nullable) = false];
> ```

i think it might be best to make the name even more specific: can we call it `generated_always`

then the comment about `OVERRIDING SYSTEM VALUE` will be more accurate, since `generated always` is the only case where that matters
___
*[pkg/sql/catalog/tabledesc/column.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg76bEf9AwkPTcKGwOL:-Mg76bEf9AwkPTcKGwOM:biyge11) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/tabledesc/column.go#L192)):*
> ```Go
> // IsDisallowedForExplicitWrite returns true iff the column is
> // not allowed for explicit write (write without OVERRIDING SYSTEM VALUE token)
> func (w column) IsDisallowedForExplicitWrite() bool {
> ```

all the other names/place where we use `IsDisallowedForExplicitWrite` should also be renamed to `IsGeneratedAlwaysColumn`
___
*[pkg/sql/pgwire/pgcode/codes.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Ccl-6NYD1o7qCf4:-Mg75Ccl-6NYD1o7qCf5:bioptgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/pgwire/pgcode/codes.go#L136)):*
> ```Go
> 	ForeignKeyViolation          = MakeCode("23503")
> 	UniqueViolation              = MakeCode("23505")
> 	DisallowExplicitWrite        = MakeCode("23506")
> ```

we need to be careful adding codes here. these actually need to match with the Postgres error codes: https://www.postgresql.org/docs/13/errcodes-appendix.html

It looks like there is no `23506` code. when i tried with postgres, it looks like it uses a `428C9` (named `generated_always`) for this type of error, so let's use that code
___
*[pkg/sql/sqlerrors/errors.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg76vq11yDhPyrEvh8V:-Mg76vq11yDhPyrEvh8W:bn7tx5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/sqlerrors/errors.go#L54)):*
> ```Go
> // NewDisallowExplicitWriteError creates an error for a violation of a disallow explicit write constraint.
> func NewDisallowExplicitWriteError(columnName string) error {
> 	return pgerror.Newf(pgcode.DisallowExplicitWrite, "column %q cannot be explicitly overridden "+
> ```

let's match the Postgres error a bit more. Postgres shows this:

```
cannot insert into column "b"
DETAIL:  Column "b" is an identity column defined as GENERATED ALWAYS.
HINT:  Use OVERRIDING SYSTEM VALUE to override.
```

so i think we can do
```
	return errors.WithHint(
		errors.WithDetailf(
			pgerror.Newf(pgcode.GeneratedAlways, "cannot insert into column %q", columnName),
			"Column %q is an identity column defined as GENERATED ALWAYS.",
			columnName,
		),
		"Use OVERRIDING SYSTEM VALUE to override.",
	)
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
looks good from my end; but just some small nits. once those are addressed, and after Marcus has a chance to review this, this should be good to go

i see that this includes the syntax changes from https://github.com/cockroachdb/cockroach/pull/67956 now. i think we should order the commits as follows:

```
sql/parser: support GENERATED ALWAYS / BY DEFAULT AS IDENTITY syntax
Add GeneratedAlways bool attribute to column definition
Reorganize sql.processSerialInColumnDef method into helper functions 
Add logictest for GENERATED AS IDENTITY syntax under CREATE TABLE (can also be squashed into previous commit)
```

Also for all the commit messages, make add the prefix `sql: ` to the commit message title.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-Mg7b0Ty2-lpjypKFjjD:b-fhfao7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1336 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7avxq42zucIJm9Y6G:-Mg7avxq42zucIJm9Y6H:bnpxfr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/opt/optbuilder/mutation_builder.go#L1336)):*
> ```Go
> //
> // TODO(janexing): to implement the OVERRIDING SYSTEM VALUE syntax
> // under `INSERT/UPSERT/UPDATE` statement
> ```

nit: add a link to https://github.com/cockroachdb/cockroach/issues/68201
___
*[pkg/sql/pgwire/pgcode/codes.go, line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7atanDYpfHGZgC2sm:-Mg7atanDYpfHGZgC2sn:b-f29wxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/pgwire/pgcode/codes.go#L213)):*
> ```Go
> 	IndeterminateCollation             = MakeCode("42P22")
> 	WrongObjectType                    = MakeCode("42809")
> 	GeneratedAlways										 = MakeCode("428C9")
> ```

nit: fix whitespace
___
*[pkg/sql/sqlerrors/errors.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7aeQO7k3Ond2vDlM8:-Mg7aeQP6yk753rWAbZd:byjfe4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/sqlerrors/errors.go#L60)):*
> ```Go
> 			"Column %q is an identity column defined as GENERATED ALWAYS.", columnName,
> 		),
> 		"Use OVERRIDING SYSTEM VALUE to override.",
> ```

oh, sorry! i wasn't thinking in my earlier comment. let's not add the OVERRIDING SYSTEM VALUE hint, since we don't support that syntax yet.

there can be a comment to add a "Use OVERRIDING SYSTEM VALUE to override." hint once https://github.com/cockroachdb/cockroach/issues/68201 is resolved


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7eJB24IXp208Eoiic:-Mg7eJB24IXp208Eoiid:b-jos4av) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/descpb/structured.proto#L156)):*
> ```protobuf
>   // To insert a generated_always column, the INSERT statement
>   // must include the OVERRIDING SYSTEM VALUE clause
>   optional bool generated_always = 18 [(gogoproto.nullable) = false];
> ```

actually, i just realized one more thing. we might need to distinguish between generated_by_default columns as well

for example, in the pg_catalog, there is a `attidentity` column that needs to be `a` for "always" for `d` for "default`. see https://www.postgresql.org/docs/13/catalog-pg-attribute.html

so i believe this field should actually be an optional enum. sorry i didn't remember that sooner


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
I think there is a misunderstanding about the behavior of `GENERATED ALWAYS AS IDENTITY`. From the documentation:

> This clause creates the column as an identity column. It will have an implicit sequence attached to it and **the column in new rows will automatically have values from the sequence assigned to it**. Such a column is implicitly NOT NULL.

So the `GENERATED ALWAYS` part signifies that the column cannot be written to (without the override), but the `AS IDENTITY` part is indicating that the value at insert is generated from a sequence. For example:

```
marcus=# CREATE TABLE t (a INT, b INT GENERATED ALWAYS AS IDENTITY);
CREATE TABLE

marcus=# INSERT INTO t VALUES (1);
INSERT 0 1

marcus=# INSERT INTO t VALUES (2);
INSERT 0 1

marcus=# INSERT INTO t VALUES (30);
INSERT 0 1

marcus=# SELECT * FROM t;
 a  | b
----+---
  1 | 1
  2 | 2
 30 | 3
(3 rows)
```

We'll also want to add tests for this feature, including logic tests and optimizer-specific tests (which I can give some direction on once we get the core semantics of `GENEREATE ALWAYS AS IDENTITY` figured out.

And a minor nit: we try to wrap git commit messages at 100 characters (I usually do 80) to make `git log` mroe readable on narrow terminals. More git commit guidelines can be found here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#GitCommitMessages-Commitdescription

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-Mg7bQ67BLKZGRLVZL6J:b-ysmi05)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Z3q9LVcgDfht2S6:-Mg7_c2B1rT0NqPzRvpC:b-v204k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L154)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think it might be best to make the name even more specific: can we call it `generated_always`

then the comment about `OVERRIDING SYSTEM VALUE` will be more accurate, since `generated always` is the only case where that matters
</blockquote></details>

The keywords `GENERATED ALWAYS` are also used to create computed columns (see example below), so I find this name confusing. What about `identity` instead? Another idea: the name can be something that makes sense to us, even if it's not part of the keywords that set it.

```
marcus=# CREATE TABLE c (a INT, b INT GENERATED ALWAYS AS (a + 1) STORED);
CREATE TABLE

marcus=# \d c
                               Table "public.c"
 Column |  Type   | Collation | Nullable |              Default
--------+---------+-----------+----------+------------------------------------
 a      | integer |           |          |
 b      | integer |           |          | generated always as (a + 1) stored
```
___
*[pkg/sql/catalog/tabledesc/table.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7etpp4Dsjxxd8J-Wc:-Mg7etpp4Dsjxxd8J-Wd:b32vno) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/tabledesc/table.go#L73)):*
> ```Go
> 		Virtual:               d.IsVirtual(),
> 		Hidden:                d.Hidden,
> 		GeneratedAlways:       d.Generated.IsGenerated && d.Generated.GeneratedType == tree.GeneratedAlways,
> ```

A 


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/catalog/tabledesc/table.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7etpp4Dsjxxd8J-Wc:-Mg7fyX312-ZxKjC-g9o:btf83nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/tabledesc/table.go#L73)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

A 
</blockquote></details>

My comment was lost, sorry.

I don't see any parsing changes, so I don't know what the AST nodes are for `GENERATED ALWAYS`, but I think we only want to set this bool if the syntax was `GENERATED ALWAYS AS IDENTITY`, not just for any `GENERATED ALWAYS` keywords.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
@mgartner: Jane had another PR that adds the syntax https://github.com/cockroachdb/cockroach/pull/67956. i also left a comment about how we'll want to make sure that these related commits go in in the right order. that PR also has basic tests in it, but we'll want more.

> So the GENERATED ALWAYS part signifies that the column cannot be written to (without the override), but the AS IDENTITY part is indicating that the value at insert is generated from a sequence.

I think we're all on the same page about this. It just wasn't obvious since there was a different PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-Mg7hD6Y81afl_6SZS3W:b-usgfq5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Z3q9LVcgDfht2S6:-Mg7habh5urcRN102TIv:b-40m7y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L154)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

The keywords `GENERATED ALWAYS` are also used to create computed columns (see example below), so I find this name confusing. What about `identity` instead? Another idea: the name can be something that makes sense to us, even if it's not part of the keywords that set it.

```
marcus=# CREATE TABLE c (a INT, b INT GENERATED ALWAYS AS (a + 1) STORED);
CREATE TABLE

marcus=# \d c
                               Table "public.c"
 Column |  Type   | Collation | Nullable |              Default
--------+---------+-----------+----------+------------------------------------
 a      | integer |           |          |
 b      | integer |           |          | generated always as (a + 1) stored
```

</blockquote></details>

as per my comment on r2, i think we actually want this to be an enum

i think the name `identity` makes sense, and the enum values can be `generated_always` or `generated_by_default`


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/catalog/tabledesc/table.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7etpp4Dsjxxd8J-Wc:-Mg7ky5D2qArv4f2e5AT:bypqtlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/tabledesc/table.go#L73)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

My comment was lost, sorry.

I don't see any parsing changes, so I don't know what the AST nodes are for `GENERATED ALWAYS`, but I think we only want to set this bool if the syntax was `GENERATED ALWAYS AS IDENTITY`, not just for any `GENERATED ALWAYS` keywords.
</blockquote></details>

@mgartner Thanks for reviewing! The parser support is implementing in another PR's [tree/create.go](https://github.com/cockroachdb/cockroach/pull/67956/commits/80901e3fde69686a2cbc3f9e90c8618cb09dbc68#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67) and [tree/sql.y](https://github.com/cockroachdb/cockroach/pull/67956/commits/80901e3fde69686a2cbc3f9e90c8618cb09dbc68#diff-570c32ab78bd63a6e239714da4bcbaed11c42a87d8b0f44552a646875ba9b317). And yes, we set this bool only when there is `GENERATED ALWAYS AS IDENTITY` token in the statement.  Sorry for the confusion!


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/catalog/tabledesc/table.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7etpp4Dsjxxd8J-Wc:-Mg7nmI13ZKReKp9_bJC:bt4l8xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/tabledesc/table.go#L73)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

@mgartner Thanks for reviewing! The parser support is implementing in another PR's [tree/create.go](https://github.com/cockroachdb/cockroach/pull/67956/commits/80901e3fde69686a2cbc3f9e90c8618cb09dbc68#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67) and [tree/sql.y](https://github.com/cockroachdb/cockroach/pull/67956/commits/80901e3fde69686a2cbc3f9e90c8618cb09dbc68#diff-570c32ab78bd63a6e239714da4bcbaed11c42a87d8b0f44552a646875ba9b317). And yes, we set this bool only when there is `GENERATED ALWAYS AS IDENTITY` token in the statement.  Sorry for the confusion!
</blockquote></details>

Ohhhh I see. Sorry about that! I'll take a look at that PR to familiarize myself with those changes.

As tests are added here I think it will make it more clear what this PR accomplishes. I'd also suggest making it clear in the commit message what is and isn't included for the feature. The statement "This matches the PostgreSQL syntax:" in the commit message made me think this commit was supposed to be the feature in its entirety.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/catalog/tabledesc/table.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7etpp4Dsjxxd8J-Wc:-Mg7ole8BeVyDFwaZPAB:b-bv1ew3) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/tabledesc/table.go#L73)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Ohhhh I see. Sorry about that! I'll take a look at that PR to familiarize myself with those changes.

As tests are added here I think it will make it more clear what this PR accomplishes. I'd also suggest making it clear in the commit message what is and isn't included for the feature. The statement "This matches the PostgreSQL syntax:" in the commit message made me think this commit was supposed to be the feature in its entirety.
</blockquote></details>

Yeah definitely, I'll add this to the commit message. Thanks for pointing out!


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/catalog/descpb/structured.proto, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7eJB24IXp208Eoiic:-Mg7p8eW4rSQT9-UR_VK:b-160w6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/descpb/structured.proto#L156)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

actually, i just realized one more thing. we might need to distinguish between generated_by_default columns as well

for example, in the pg_catalog, there is a `attidentity` column that needs to be `a` for "always" for `d` for "default`. see https://www.postgresql.org/docs/13/catalog-pg-attribute.html

so i believe this field should actually be an optional enum. sorry i didn't remember that sooner
</blockquote></details>

Maybe this is not necessary? Since we treat the `generated_by_default` column the same as a `SERIAL` column (with `serial_normalization == sql_sequence`). 

We add this `generated_always` bool to `ColumnDescriptor` because `INSERT` needs to check it.  But for `generated_by_default` there's no need for such a checking.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/catalog/descpb/structured.proto, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7eJB24IXp208Eoiic:-Mg7qh1dCxFSk3iG79sY:b-2gfvm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/descpb/structured.proto#L156)):*
Oops sorry, I misunderstood.

> i think the name `identity` makes sense, and the enum values can be `generated_always` or `generated_by_default`

Yeah I think it makes much sense to have it as enum instead if two bools, will do!


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> i think we should order the commits as follows:
> sql/parser: support GENERATED ALWAYS / BY DEFAULT AS IDENTITY syntax
> Add GeneratedAlways bool attribute to column definition
> Reorganize sql.processSerialInColumnDef method into helper functions 
> Add logictest for GENERATED AS IDENTITY syntax under CREATE TABLE (can also be squashed into previous commit)

This makes sense to me because it allows each commit to have tests for the incremental changes.

How does https://github.com/cockroachdb/cockroach/pull/67956 fit into this? Is the plan to have all the commits in a single PR or split them into multiple PRs?

I typically put larger features behind feature flags until they are complete, so that, for example, if the feature is not fully functional by 21.2.0-alpha.1 (will be released on the Aug 24 nightly) won't introduce something that is partially working. There's no clear line for when a feature is big enough to do this, so it's up to you. Here's an example of such a flag being added: https://github.com/cockroachdb/cockroach/commit/68f10e5a8a188f2783f576961518fa45260eab06

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-MgGgDA0A_zH2px-iEfV:bo9vwtz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Z3q9LVcgDfht2S6:-Mg7gLqb4INJf6pwyzDP:bw2bque) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L154)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

The keywords `GENERATED ALWAYS` are also used to create computed columns (see example below), so I find this name confusing. What about `identity` instead? Another idea: the name can be something that makes sense to us, even if it's not part of the keywords that set it.

```
marcus=# CREATE TABLE c (a INT, b INT GENERATED ALWAYS AS (a + 1) STORED);
CREATE TABLE

marcus=# \d c
                               Table "public.c"
 Column |  Type   | Collation | Nullable |              Default
--------+---------+-----------+----------+------------------------------------
 a      | integer |           |          |
 b      | integer |           |          | generated always as (a + 1) stored
```

</blockquote></details>

@rafiss suggestion for an enum is a good one too. My main concern is avoiding confusion with computed columns.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
just some smaller comments this time! nice work

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-MgHgfbR1XRWilmXeXx8:b6rxqjt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 199 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHewztBOUnmeGzfVGL:-MgHewztBOUnmeGzfVGM:bpmmhi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/catalog/descpb/structured.proto#L199)):*
> ```protobuf
> enum GeneratedAsIdentityType {
>   // A column created without `GENERATED ... AS IDENTITY` token.
>   NOTIDENTITYCOLUMN = 0;
> ```

nit: these enum names can have underscores to improve readability. i'm not sure why the one above this doesn't do that, but no need to match that style.
___
*[pkg/sql/opt/cat/column.go, line 187 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHfH9_97P6w2qess8h:-MgHfH9a1Fb-T84D8whH:bmliapy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/opt/cat/column.go#L187)):*
> ```Go
> // is associated with the GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY
> // token.
> type GeneratedAsIdentityType uint8
> ```

not sure about this part of the optimizer, but does this need to be duplicated here? we can't use the protobuf enum?

assuming it does need to be duplicated, do the enum values need to match? if so, then i'd recommend hardcoding the values rather than using `iota`
___
*[pkg/sql/parser/sql.y, line 6165 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHftPR5kQ_-MSzaSVL:-MgHftPR5kQ_-MSzaSVM:b-wn97u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6165)):*
> ```Yacc
> //
> // Column qualifiers:
> //   [CONSTRAINT <constraintname>] {NULL | NOT NULL | NOT VISIBLE | UNIQUE | PRIMARY KEY | CHECK (<expr>) | DEFAULT <expr> | GENERATED { ALWAYS | BY DEFAULT } AS IDENTITY [( <opt_sequence_option_list> )]}
> ```

this doesn't look like the right place to add the syntax to the docs -- this line is referring to column constraints.

instead, `GENERATED ... IDENTITY` should go on a separate line after `AS ( <expr> ) STORED`
___
*[pkg/sql/sem/tree/pretty.go, line 1899 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHgRUi3seV8hHFcNIE:-MgHgRUi3seV8hHFcNIF:bxyyat2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/sem/tree/pretty.go#L1899)):*
> ```Go
> 	// GENERATED ALWAYS/BY DEFAULT AS IDENTITY constraint.
> 	if node.GeneratedIdentity.IsGenerated {
> 		generatedConstraint := pretty.Nil
> ```

see above; the `GENERATED ... IDENTITY` should not show up as a column constraint. i don't believe we need this code at all


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> Reviewed 7 of 24 files at r4, 8 of 17 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/table.go, line 76 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgIZYgI27IXgDUjbp3U:-MgIZYgI27IXgDUjbp3V:b-w56xjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/catalog/tabledesc/table.go#L76)):*
> ```Go
> 
> 	if d.GeneratedIdentity.IsGenerated && d.GeneratedIdentity.GeneratedType == tree.GeneratedAlways {
> 		col.GeneratedAsIdType = descpb.GeneratedAsIdentityType_GENERATEDALWAYS
> ```

We should be assigning `col.GeneratedAsIdType` GENERATED DEFAULT here as well right?
___
*[pkg/sql/opt/cat/column.go, line 187 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHfH9_97P6w2qess8h:-MgIae-7BEEYvi3474rX:b-z28rvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/opt/cat/column.go#L187)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

not sure about this part of the optimizer, but does this need to be duplicated here? we can't use the protobuf enum?

assuming it does need to be duplicated, do the enum values need to match? if so, then i'd recommend hardcoding the values rather than using `iota`
</blockquote></details>

@rafiss don't think we can import the `enum` definitions defined in the protobuf / descpb here but I think we can define the enum as an int elsewhere and then cast the field in the protobuf to the go enum.
___
*[pkg/sql/parser/sql.y, line 6723 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgIa-yVDI8gQzu0kvWX:-MgIa-yVDI8gQzu0kvWY:b-obzb71) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6723)):*
> ```Yacc
>   }
> | generated_by_default_as IDENTITY
> 	{
> ```

nit: spacing doesn't match up here, with the previous entry.
Recommend using spaces instead of tabs in sql.y
___
*[pkg/sql/parser/sql.y, line 6744 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgIa5NXEsZRefSWm7gS:-MgIa5NXEsZRefSWm7gT:b-b0l55j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6744)):*
> ```Yacc
> // GENERATED ALWAYS is a noise word for compatibility with Postgres.
> generated_always_as:
> 	GENERATED_ALWAYS ALWAYS AS {}
> ```

nit: use spaces instead of tabs


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/descpb/structured.proto, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Z3q9LVcgDfht2S6:-MgLrM2l6_yRGL3Ibr9-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/descpb/structured.proto#L154)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

@rafiss suggestion for an enum is a good one too. My main concern is avoiding confusion with computed columns.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descpb/structured.proto, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7eJB24IXp208Eoiic:-MgLfeSd8q_oA6WefJhZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/descpb/structured.proto#L156)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Oops sorry, I misunderstood.

> i think the name `identity` makes sense, and the enum values can be `generated_always` or `generated_by_default`

Yeah I think it makes much sense to have it as enum instead if two bools, will do!
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/column.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg76bEf9AwkPTcKGwOL:-MgLfbdF344poomagurY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/catalog/tabledesc/column.go#L192)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

all the other names/place where we use `IsDisallowedForExplicitWrite` should also be renamed to `IsGeneratedAlwaysColumn`
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/table.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7etpp4Dsjxxd8J-Wc:-MgLqP4J5uIF3VXyEMvW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/catalog/tabledesc/table.go#L73)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Yeah definitely, I'll add this to the commit message. Thanks for pointing out!
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/table.go, line 76 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgIZYgI27IXgDUjbp3U:-MgLqPnX5zUBVkIlVKGG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/catalog/tabledesc/table.go#L76)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

We should be assigning `col.GeneratedAsIdType` GENERATED DEFAULT here as well right?
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/column.go, line 187 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHfH9_97P6w2qess8h:-MgLpPR70x55HFb5F5Sm:b58jduk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/opt/cat/column.go#L187)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

@rafiss don't think we can import the `enum` definitions defined in the protobuf / descpb here but I think we can define the enum as an int elsewhere and then cast the field in the protobuf to the go enum.
</blockquote></details>

@richardjcai @rafiss Importing `enum` from `descpb` seems working. Just made [the change](https://github.com/cockroachdb/cockroach/pull/68324/commits/0e420ac7b75217ac5b47c35167799b58911dbbfd#diff-6924e22ddd574685bd25123a624dd286abfae52928e4a74d3dbfbc3bcc0411c6R39). 
___
*[pkg/sql/parser/sql.y, line 6165 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgHftPR5kQ_-MSzaSVL:-MgLqepi3uDKjW7dvUIZ:bbk7t9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6165)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this doesn't look like the right place to add the syntax to the docs -- this line is referring to column constraints.

instead, `GENERATED ... IDENTITY` should go on a separate line after `AS ( <expr> ) STORED`
</blockquote></details>

@rafiss According to the [latest PostgreSQL document for `CREATE TABLE`](https://www.postgresql.org/docs/current/sql-createtable.html), it seems that `GENERATED ... IDENTITY` is on the same level as other column constraints such as `NOT NULL` and `UNIQUE`.
___
*[pkg/sql/pgwire/pgcode/codes.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg75Ccl-6NYD1o7qCf4:-MgLrHtwDmMlWjqAthrT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/pgwire/pgcode/codes.go#L136)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we need to be careful adding codes here. these actually need to match with the Postgres error codes: https://www.postgresql.org/docs/13/errcodes-appendix.html

It looks like there is no `23506` code. when i tried with postgres, it looks like it uses a `428C9` (named `generated_always`) for this type of error, so let's use that code
</blockquote></details>

Done.
___
*[pkg/sql/sqlerrors/errors.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg76vq11yDhPyrEvh8V:-MgLf9jm1vELOXUKBBtz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f78aa9fa9409c8f8864982fc8e8621ad6b2357b3/pkg/sql/sqlerrors/errors.go#L54)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

let's match the Postgres error a bit more. Postgres shows this:

```
cannot insert into column "b"
DETAIL:  Column "b" is an identity column defined as GENERATED ALWAYS.
HINT:  Use OVERRIDING SYSTEM VALUE to override.
```

so i think we can do
```
	return errors.WithHint(
		errors.WithDetailf(
			pgerror.Newf(pgcode.GeneratedAlways, "cannot insert into column %q", columnName),
			"Column %q is an identity column defined as GENERATED ALWAYS.",
			columnName,
		),
		"Use OVERRIDING SYSTEM VALUE to override.",
	)
```

</blockquote></details>

Done.
___
*[pkg/sql/sqlerrors/errors.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mg7aeQO7k3Ond2vDlM8:-MgLfARg4dy21DLm254I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/055f62230147ae3aa9d23db5b71f70f3a5ccce09/pkg/sql/sqlerrors/errors.go#L60)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

oh, sorry! i wasn't thinking in my earlier comment. let's not add the OVERRIDING SYSTEM VALUE hint, since we don't support that syntax yet.

there can be a comment to add a "Use OVERRIDING SYSTEM VALUE to override." hint once https://github.com/cockroachdb/cockroach/issues/68201 is resolved
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
this looks almost ready to go. looks like you need to rebase with `git pull --rebase origin mater`

and fix `TestDoctorZipDir/recreate` and `TestExecBuild/local/show_trace_nonmetamorphic`and `TestValidateCoversAllDescriptorFields`

i believe for the first two you can do
```
make test PKG=./pkg/cli TESTS='TestDoctorZipDir' TESTFLAGS='-rewrite'
make test ptlogic FILES='show_trace_nonmetamorphic' TESTFLAGS='-rewrite'
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-MgMvqur7abx3ncdkB44:bj57dm9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/sem/tree/pretty.go, line 1839 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgMuhzt08EFXG67aaPS:-MgMuhzt08EFXG67aaPT:bghrk6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/64ad55bdac709d6b1b74c301f090dc34034dfc5b/pkg/sql/sem/tree/pretty.go#L1839)):*
> ```Go
> 	//   [[CREATE [IF NOT EXISTS]] FAMILY [name]]
> 	//   [[CONSTRAINT name] DEFAULT expr]
> 	//   GENERATED {ALWAYS|BY DEFAULT} AS IDENTITY
> ```

nit: can you move this line so it comes after `[AS ... STORED]`


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql/parser: support GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY syntax

This commit is to allow the parser to support
the `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` syntax
under `CREATE TABLE` statement.

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release note (sql change): Added support for `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY`
syntax in a column definition. This will automatically create
a sequence for the given column.
This matches PostgreSQL syntax and functionality.
See https://www.postgresql.org/docs/current/sql-createtable.html

Release Note: None

ZhouXing19 committed:
sql: add GeneratedAsIdentityType enum attribute to column definition

This commit add a enum attribute to note how the column creation is associated
with the `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` token.
If a column is created with `GENERATED ALWAYS AS IDENTITY`, it is not allowed
to be overridden without `OVERRIDING SYSTEM VALUE` token in the `INSERT
| UPSERT | UPDATE` statement.
See also #68201

We implement `GeneratedAsIdentityType` enum following this logic:

1. Persistent column definition: Add `enum GeneratedAsIdentityType`, and
add `GeneratedAsIdentityType generated_as_id_type` under `ColumnDescriptor`
at `pkg/sql/catalog/descpb/structured.proto`

2. Write access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
set the variable when `GENERATED ALWAYS AS IDENTITY` is in the
`CREATE TABLE` statement. Check `pkg/sql/catalog/tabledesc/table.go`

3. Read access to `descpb.ColumnDescriptor.GeneratedAsIdType`:
refactored `cat.Column.Init()` at `pkg/sql/opt/cat/column.go`

4. Check if a column is created with `GENERATED ALWAYS AS IDENTITY` when
insert: add `func checkColumnIsGeneratedAlways`
at `pkg/sql/opt/optbuilder/mutation_builder.go`
and implement it in `pkg/sql/opt/optbuilder/insert.go`
and `pkg/sql/opt/optbuilder/insert.go`
and `pkg/sql/opt/optbuilder/update.go`

We also added logic tests at
`pkg/sql/logictest/testdata/logic_test/alter_table`
and `pkg/sql/logictest/testdata/logic_test/create_table`

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release note: None

rafiss commented on pkg/sql/parser/testdata/create_table:
```diff
@@ -2260,3 +2260,15 @@ CREATE TABLE operator_tbl (a INT8 DEFAULT 1 OPERATOR(+) 2, b INT8 DEFAULT 1 OPER
+CREATE TABLE generated_always_as_identity_tb (
+  a INT UNIQUE,
+  b INT GENERATED ALWAYS AS IDENTITY,
+  c INT GENERATED BY DEFAULT AS IDENTITY
```
can you also add a test to make sure it doesn't work if you try multiple GENERATED types?

for example, these shouldn't work (for both CREATE and ALTER TABLE):
```
CREATE TABLE generated_error (
   a INT UNIQUE,
   b INT GENERATED BY DEFAULT AS IDENTITY GENERATED ALWAYS AS (a) STORED
 )
-- postgres returns: both identity and generation expression specified for column "b" of table "generated_error"

CREATE TABLE generated_error (
   a INT UNIQUE,
   b INT GENERATED BY DEFAULT AS IDENTITY GENERATED ALWAYS AS IDENTITY
 )
-- postgres returns: multiple identity specifications for column "b" of table "generated_error"
```

see https://github.com/cockroachdb/cockroach/blob/6fafc569ad64c70654345dd6c685c6b32991307b/pkg/sql/parser/testdata/alter_default_privileges#L49 for an example of an error test

rafiss commented on pkg/sql/catalog/tabledesc/validate_test.go:
```diff
@@ -175,8 +175,9 @@ var validationMap = []struct {
 				reason: "initial import: TODO(features): add validation"},
-			"Hidden":       {status: iSolemnlySwearThisFieldIsValidated},
-			"Inaccessible": {status: iSolemnlySwearThisFieldIsValidated},
+			"GeneratedAsIdType": {status: iSolemnlySwearThisFieldIsValidated},
```
was the field actually validated? :) 

you should add something here: https://github.com/cockroachdb/cockroach/blob/6fafc569ad64c70654345dd6c685c6b32991307b/pkg/sql/catalog/tabledesc/validate.go#L685

it should make sure the descriptor is not an an incoherent state. e.g., a `computed` column cannot be a `generatedAsIdentity` column.

ZhouXing19 replied:
@rafiss  Done, check [here](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-3feb8ec23c24149443848d5a36e11d9dd93cfddf4a95363a80c4b79605664667R706-R709). Also added [a `NOT NULL` constraint](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R544-R546), and corresponding [parser tests](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-dccba38be048680977583650bddede7895ab54cccce0fbccabcd66f2044402e8R2276-R2302) and [logic tests](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-fa7c476e990651baf64186090334103fb4acfd5ff9b37e6f699be44e27d61e61R493).

rafiss reviewed (approved):
after fixing this, it looks good from my end! (after mgartner's sign off too)

mgartner reviewed (commented):
Nice work! I left a few more minor comments, but this is getting really close!

nit: the PR description is copied to the merge commit once the PR is merged with bors, so ideally lines are wrapped at 80 characters just like your commit messages. I usually copy the commit message(s) directly into the PR description.

> Reviewed 4 of 24 files at r4, 3 of 23 files at r10, 3 of 23 files at r12, 21 of 21 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-Mgg7Wba9aVFDa7taxAW:bp2n6qc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/parser/sql.y, line 6725 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg3zAB9t0lYhs6J9qI:-Mgg3zAB9t0lYhs6J9qJ:bpbaugv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6725)):*
> ```Yacc
> 	{
> 	   $$.val = &tree.GeneratedByDefAsIdentity{}
> 	}
> ```

nit: there's still 2 lines with tabs instead of spaces
___
*[pkg/sql/parser/sql.y, line 6742 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg42Fs0naQOBkBhJu_:-Mgg42Fs0naQOBkBhJua:b-n11xmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6742)):*
> ```Yacc
> | generated_always_as
> 
> // GENERATED ALWAYS is a noise word for compatibility with Postgres.
> ```

nit: From what I can tell, "noise word" meant that this syntax had no semantic meaning. Now that it does, it's no longer a "noise word". You should update this comment, or just remove it.
___
*[pkg/sql/parser/sql.y, line 6746 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg4GRB2hIvuUAEXvHp:-Mgg4GRB2hIvuUAEXvHq:bktv5va) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6746)):*
> ```Yacc
> 	GENERATED_ALWAYS ALWAYS AS {}
> 
> // GENERATED BY DEFAULT is a noise word for compatibility with Postgres.
> ```

nit: it's no longer a noise word
___
*[pkg/sql/parser/testdata/alter_table, line 1109 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg597I4_K-DHrCnglf:-Mgg597I4_K-DHrCnglg:b-rbh4q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/parser/testdata/alter_table#L1109)):*
> ```
> DETAIL: source SQL:
> ALTER TABLE a ADD COLUMN b INT GENERATED BY DEFAULT AS IDENTITY GENERATED ALWAYS AS IDENTITY
>                                                                                             ^
> ```

nit: add error cases for computed columns, like I mention below in `CREATE TABLE`
___
*[pkg/sql/parser/testdata/create_table, line 2302 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg4jbcCgXkRXvggjKN:-Mgg4jbd-P93nWhvNfLV:bvhrfr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/parser/testdata/create_table#L2302)):*
> ```
>   b INT NULL GENERATED BY DEFAULT AS IDENTITY
> )
> ^
> ```

nit: add a cases for a computed columns and this new syntax, like Rafi suggested:

```
CREATE TABLE generated_error (
  a INT,
  b INT AS (a + 10) STORED GENERATED ALWAYS AS IDENTITY
)
```

```
CREATE TABLE generated_error (
  a INT,
  b INT GENERATED ALWAYS AS (a + 10) STORED GENERATED ALWAYS AS IDENTITY
)
```

A case like this would be nice too:

```
CREATE TABLE generated_error (
  a INT,
  b INT GENERATED ALWAYS AS (a + 10) IDENTITY
)
```
___
*[pkg/sql/sem/tree/create.go, line 408 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg5OYs9gMceoeYhc3y:-Mgg5OYs9gMceoeYhc3z:b-7edlqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/sem/tree/create.go#L408)):*
> ```Go
> // Generatedtype represents either GENERATED ALWAYS AS IDENTITY
> // or GENERATED BY DEFAULT AS IDENTITY
> type Generatedtype int
> ```

nit: `Generatedtype` => `GeneratedType`

Maybe it should be named `GeneratedIdentityType`, since it only applies with `GENERATED ... AS IDENTITY`?
___
*[pkg/sql/sem/tree/create.go, line 423 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg60cO6dh1OeKI2DWy:-Mgg60cO6dh1OeKI2DWz:b-pkfiup) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/sem/tree/create.go#L423)):*
> ```Go
> 	IsSerial          bool
> 	GeneratedIdentity struct {
> 		IsGenerated   bool
> ```

nit: Maybe a better name is `IsGeneratedIdentity` to avoid confusion with the `GENERATED ALWAYS AS (a + 1) STORED` alternative syntax for computed columns?
___
*[pkg/sql/sem/tree/pretty.go, line 1877 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg71g-AaPe8yUsgaTp:-Mgg71g-AaPe8yUsgaTq:b-u0qidk) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/sem/tree/pretty.go#L1877)):*
> ```Go
> 	// GENERATED ALWAYS/BY DEFAULT AS IDENTITY constraint.
> 	if node.GeneratedIdentity.IsGenerated {
> 		generatedConstraint := pretty.Nil
> ```

nit: `var generatedConstraint pretty.Doc` is more idiomatic, I think.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/parser/sql.y, line 6725 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg3zAB9t0lYhs6J9qI:-MggKGE16SFWXqzw8s_L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6725)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: there's still 2 lines with tabs instead of spaces
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 6742 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg42Fs0naQOBkBhJu_:-MggKHC-8UoN4tUxdJnI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6742)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: From what I can tell, "noise word" meant that this syntax had no semantic meaning. Now that it does, it's no longer a "noise word". You should update this comment, or just remove it.
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 6746 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg4GRB2hIvuUAEXvHp:-MggKF6G02eT0FFGHAkU:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6afebf2052ddf7c32a387d9eb31446e19d673244/pkg/sql/parser/sql.y#L6746)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: it's no longer a noise word
</blockquote></details>

Done
___
*[pkg/sql/parser/testdata/alter_table, line 1109 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg597I4_K-DHrCnglf:-MggK2hh-_lvLyqHRiNg:bvxt3dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/parser/testdata/alter_table#L1109)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add error cases for computed columns, like I mention below in `CREATE TABLE`
</blockquote></details>

Done, check [here](https://github.com/cockroachdb/cockroach/pull/68324/commits/a4b1083b306f1bc2911784202267758e195523ea#diff-727d81af299318702bad5a0995f002fda99360e94d6701dc38f6fb305d3aba95R1111-R1133). Thanks!
___
*[pkg/sql/parser/testdata/create_table, line 2302 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg4jbcCgXkRXvggjKN:-MggJNRe9_9A8Nh-Qfjv:b-47w3vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/parser/testdata/create_table#L2302)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a cases for a computed columns and this new syntax, like Rafi suggested:

```
CREATE TABLE generated_error (
  a INT,
  b INT AS (a + 10) STORED GENERATED ALWAYS AS IDENTITY
)
```

```
CREATE TABLE generated_error (
  a INT,
  b INT GENERATED ALWAYS AS (a + 10) STORED GENERATED ALWAYS AS IDENTITY
)
```

A case like this would be nice too:

```
CREATE TABLE generated_error (
  a INT,
  b INT GENERATED ALWAYS AS (a + 10) IDENTITY
)
```
</blockquote></details>

Done. Check [here](https://github.com/cockroachdb/cockroach/pull/68324/commits/a4b1083b306f1bc2911784202267758e195523ea#diff-dccba38be048680977583650bddede7895ab54cccce0fbccabcd66f2044402e8R2304-R2343). Thanks!
___
*[pkg/sql/sem/tree/create.go, line 408 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg5OYs9gMceoeYhc3y:-MggJHDK8zrmOvF4wdE5:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/sem/tree/create.go#L408)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `Generatedtype` => `GeneratedType`

Maybe it should be named `GeneratedIdentityType`, since it only applies with `GENERATED ... AS IDENTITY`?
</blockquote></details>

Done
___
*[pkg/sql/sem/tree/create.go, line 423 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg60cO6dh1OeKI2DWy:-MggJEN_7_Py9u93YBqU:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/sem/tree/create.go#L423)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Maybe a better name is `IsGeneratedIdentity` to avoid confusion with the `GENERATED ALWAYS AS (a + 1) STORED` alternative syntax for computed columns?
</blockquote></details>

Done
___
*[pkg/sql/sem/tree/pretty.go, line 1877 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-Mgg71g-AaPe8yUsgaTp:-MggJ7y3Cmd_pAcm0bFt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da5d86c6a7821a98be0ad6a9d44c53c911d66f1a/pkg/sql/sem/tree/pretty.go#L1877)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `var generatedConstraint pretty.Doc` is more idiomatic, I think.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
The first commit <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

The optimizer changes will require a few things:

* We should add support for `GENERATED [ALWAYS | BY DEFAULT] AS IDENTITY` to the test catalog. Here's a similar change that involved the same files you'll need to update: https://github.com/cockroachdb/cockroach/commit/2045ba2e86c53818b1c6cd32584583c737a93e55
* With test catalog support for these types of columns, we can write optimizer tests. Since your changes are mostly in the optbuilder, look in `pkg/sql/opt/optbuilder/testdata/` for examples of tests we should be writing. These tests ensure that the base query plan (which all optimizations are generated from) is semantically correct. We'll probably want tests for `INSERT`, `INSERT .. ON CONFLICT ... DO NOTHING`, `INSERT ... ON CONFLICT .. DO UPDATE ...`,`UPSERT`, and `UPDATE`.

> Reviewed 26 of 26 files at r16, 8 of 21 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-:-MggTxEv-JW1UX6Mgweq:b-ekexio)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/validate.go, line 709 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MggPz5y6fKBdvmLIPw8:-MggPz5y6fKBdvmLIPw9:b51p3r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2f131086ed042c26d39c3c92c68ab912a4f7b25/pkg/sql/catalog/tabledesc/validate.go#L709)):*
> ```Go
> 		if column.IsComputed() && column.IsGeneratedAsIdentity() {
> 			return fmt.Errorf("both identity and generation expression specified for column %q", column.GetName())
> 		}
> ```

:100: 

nit: Another validation you could add is one that ensures that a column is not nullable (e.g., `a INT NULL`), and be `IsGeneratedAsIdentity`.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 178 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MgN8t4H7hDE62RQ6S-A-r11-178:-MggQLdJBK570eNJAMAK:b-jylldq) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d79dbee0b928f565551a16b63951b3c24cd878/pkg/sql/catalog/tabledesc/validate_test.go#L178)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

@rafiss  Done, check [here](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-3feb8ec23c24149443848d5a36e11d9dd93cfddf4a95363a80c4b79605664667R706-R709). Also added [a `NOT NULL` constraint](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R544-R546), and corresponding [parser tests](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-dccba38be048680977583650bddede7895ab54cccce0fbccabcd66f2044402e8R2276-R2302) and [logic tests](https://github.com/cockroachdb/cockroach/pull/68324/files#diff-fa7c476e990651baf64186090334103fb4acfd5ff9b37e6f699be44e27d61e61R493).
</blockquote></details>

Good catch @rafiss! I didn't catch this when I discussed with @zhouxing19.
___
*[pkg/sql/sqlerrors/errors.go, line 54 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/68324#-MggPIdCE8SyyzXeWj3J:-MggPIdCE8SyyzXeWj3K:bpnka6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2f131086ed042c26d39c3c92c68ab912a4f7b25/pkg/sql/sqlerrors/errors.go#L54)):*
> ```Go
> // NewGeneratedAlwaysColumnError creates an error for explicitly writing a column
> // created with `GENERATED ALWAYS AS IDENTITY` token.
> // TODO(janexing): Should add a HINT with "Use OVERRIDING SYSTEM VALUE to override." once issue #68201 is resolved.
> ```

nit: wrap this comment at 80 characters


<!-- Sent from Reviewable.io -->
